Find Best Path Using Constraints

ArturoQuintana opened the pull request:
Pathfinder tests' extension to prove a proper functionality

# Pull Request Template

### :octocat: No issue related, more focus on the extension done to the Pathfinder application based on constraints.

### :chart_with_upwards_trend: No performance improvement.

### :bookmark_tabs: Inclusion of new tests and some improvements related to memory usage in the new implementation. Correction of some error that ends up on undesired behavior.

### :computer: Verification Process

None

### :page_facing_up: Release Notes

This PR includes improvements over the pathfinder application and covering a set of tests to support its functionality.

Ktmi committed:
Several changes to graph and tests.
 - Graph now allows for selection of algorithm.
 - Default graph algorithm is now networkx.all_shortest_paths
 - Updated tests to change algorithm to networkx.shortest_simple_paths
 - Fixed capitilization error in test_graph2

Commit:
5 unit tests skeleton

Ktmi committed:
More tests & a single bug fix
 - Tests for the following constraints
   + reliability
   + delay
   + bandwidth
 - Fixed missing dictionary entry for reliability

Commit:
added test_graph2.py

Commit:
added unit test skeleton

Commit:
working on test 4

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
added new graph test for self

Commit:
merged with David's changes

Commit:
completed test graph

Commit:
update test_graph3.py

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
update test_graph3.py

Commit:
added second test case

Commit:
finished unit tests

Ktmi committed:
Finished webui
 + Adds in support for constraints
 + Adds in support for flexibility

Ktmi committed:
Removed code from web ui used to help debug

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
test_graph3 now subclass of test_graph

Commit:
test_graph2 changed back to David's code

Commit:
test_graph3 class name now consistent

Commit:
test_graph3 removed redundant methods

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
fixes for test_graph3 and graph

Commit:
test_graph3 less verbose

Ktmi committed:
Merge remote-tracking branch 'upstream/master'

Ktmi committed:
New filter system, throws error for wrong type

Ktmi committed:
Added in new way to do flexibility

Ktmi committed:
Modifies rest endpoints.

Ktmi committed:
added brakes to constrained flexible

Ktmi committed:
Modifies pathfinder to reduce duplicate code

Ktmi committed:
Added test for type checking

Ktmi committed:
Final changes

Ktmi committed:
Slight speed optimization

Ktmi committed:
Merge pull request #5 from MarvinTorres/test3fixes

Fixed test_graph3.py

Ktmi committed:
Merge pull request #6 from MarvinTorres/test3fixes

fixed graph setup

Ktmi committed:
Merge pull request #7 from MarvinTorres/test3fixes

fixed test_path9

MarvinTorres committed:
updated change log

MarvinTorres committed:
set metric attribute input values to 0

MarvinTorres committed:
removed the icons next to the sliders

MarvinTorres committed:
declared var inside of loop block

MarvinTorres committed:
compacted sliders and added scrollbar

MarvinTorres committed:
removed trailing whitespaces from graph.py and main.py

removed trailing whitespaces from main.py

MarvinTorres committed:
Renamed and moved tests

 - test_graph now test_search_results
 - test_graph1, 2, 3 now test_search1_results1, 2, 3
 - fixed test_constrained_path9 in test_search_results1
 - class names changed to reflect new file names
 - moved files to unit folder

MarvinTorres committed:
Fixed padding in graph.py

MarvinTorres committed:
Cleaned up constrained_flexible_paths and added comments

 - compacted default_edge_list and flexible assignments
 - renamed some variables
 - changed loop from for to while

MarvinTorres committed:
Removed failing tests for checking

MarvinTorres committed:
Fixed formatting issues

MarvinTorres committed:
Fixed more formatting issues

MarvinTorres committed:
Minimized local variable count

 - 5 parameters instead of 6
 - 17 local vars instead of 18

MarvinTorres committed:
Fixed all formatting errors

MarvinTorres committed:
Fixed minor formatting error

MarvinTorres committed:
Changed logic to support code reuse

MarvinTorres committed:
Revamped parameter passing.

 - calling constrained flexible paths is now more user-friendly
 - test updated to reflect changes.

MarvinTorres committed:
Renamed variable flexible to depth

MarvinTorres committed:
Added accordion item for new best constrained path

MarvinTorres committed:
Continued working on new component

MarvinTorres committed:
Changed REST API parameters

MarvinTorres committed:
Implemented cmagnobarbosa's map resize fix

MarvinTorres committed:
Fixed KytosGraph import in Main

MarvinTorres committed:
Fixed logic error in get_constrained_paths_v4

 - self.v4_is_flexible[checked] was comparing strings, not boolean.
  - Will need to change code to use bools

MarvinTorres committed:
Updated get_constrained_paths_v4 to reuse bools.

MarvinTorres committed:
Implement Humberto's fixes

MarvinTorres committed:
Updated API endpoints and removed unnecessary menu
 -  "Best Constrained Path With Flexible Metrics" now "Best Constrained Path"
 -  Original "Best Constrained Path" menu removed since it was redundant
 - v3 endpoint removed
 - v4 endpoint changed to v2/best-constrained-paths
 - variable names changed to reflect removed menu

MarvinTorres committed:
Updated comments and changed endpoints for new UI

MarvinTorres committed:
Added missing endpoint changes

MarvinTorres committed:
Updated KytosGraph, TestResults and subclasses
 - KytosGraph
   - set_path_function removed.
 - TestResults
   - test_setup method removed.
   - setup method renamed to setUp to run at the start of every test
 - TestResultsSimple, TestResultsMetaData, TestResultsEdges
   - Removed explicit setup call

MarvinTorres committed:
Gave better descriptions to test subclasses.

MarvinTorres committed:
Undid KytosGraph import change in Main.

MarvinTorres committed:
Linting fixes
 - TestResults
   - removed unused import.
 - TestResultsMetadata
   - fixed two methods with same name.

MarvinTorres committed:
Fixed JSON serialization of error message

MarvinTorres committed:
Re-added methods and reverted update_links

MarvinTorres committed:
Fixed linting issue

MarvinTorres committed:
Changed inline css tags to internal css tags

Ktmi committed:
David's starting work

Switching to unit tests.

test_setup prints the list of nodes and edges.

Adds new tests

Modified test methods for more info collection.

Added in basics for constrained shortest paths

Nearly finished filters, need to define tests

Fixes error with updating default metadata

Removes _set_default_metadata method

Updated tests for constrained shortest path

identified and fixed a problem with pathing

Added a crude working implementation of flexible

Improved code efficiency for brute force method

Added in  new graph to tests

Moved new graph for into seperate test driver

MarvinTorres committed:
Updated API

MarvinTorres committed:
Update tests/unit/test_results.py

Standardized decorators

Co-authored-by: Gleyberson Andrade <gleybersonandrade@gmail.com>

MarvinTorres committed:
Updated return values of shortest_constrained_path
   - Returns 400 BAD REQUEST if user provides an illegal attribute value

hdiogenes committed:
Remove space from decorator

MarvinTorres committed:
Multiple changes to improve UX
   - Added "Minimum Flexible Hits" text to menu
   - Changed "maximum_misses" var name to "minimum_flexible_hits"

MarvinTorres committed:
Changed maximum misses variable to minimum hits
   - This is to reflect changes from #62

MarvinTorres committed:
Removed changelog update

MarvinTorres committed:
Set up mock graph with metadata

MarvinTorres committed:
Updated testing structure and files
   - test_results and children were moved to integration folder
   - test_graph and test_main now test our added methods
   - test_filter added
   - methods added to test helper

MarvinTorres committed:
Linting fixes

slinksoft committed:
Began adding the exact path prototype to graph.py

slinksoft committed:
Remove old prototype, add test implementation

Remove old prototype from graph.py, added test implementation to return node and edge data from topology built from mininet

slinksoft committed:
added improved prototype to pathfinder - ready for testing

slinksoft committed:
fix minor errors

slinksoft committed:
Update main.py

slinksoft committed:
Integrated the latest exact path algorithm into graph.py

slinksoft committed:
Integrated ExactDelayPathFinder library into Pathfinder

Also, the library will automatically install when installing this version of the Pathfinder NApp

slinksoft committed:
Change requirements to always install latest version of exactdelaypathfinder

MarvinTorres committed:
Changed main.kytos to old merge_All-Marvin version

MarvinTorres committed:
Add delay pathfinder panel and integration test

MarvinTorres committed:
Add test and change setup to run integration tests

MarvinTorres committed:
Change UI to show link and interface aliases
   - If a link has an alias, show it. Otherwise show its dpid
   - If an interface has an alias, show it. Else show its port number.
   - Show dpid if interface has neither (rare)

MarvinTorres committed:
Update UI to show exact delay results

MarvinTorres committed:
Fix "content not defined" vue js error

ArturoQuintana committed:
Minor modifications following PEP8 convention related to comment formats and whitespaces

ArturoQuintana committed:
Fixing the fact that the keys list can storage repeated elements. This might induce a memory problem that might negatively impact the scalability of the application.

ArturoQuintana committed:
Simplification of the boolean expression

ArturoQuintana committed:
Elimination of an unused variable due to a succeeding redefinition

ArturoQuintana committed:
Minor modifications following PEP8 convention related to comment formats, whitespaces, and uses of parenthesis

ArturoQuintana committed:
Refactored topology generation to externalize each process (switch creation, links, and metadata) as an independent methods

ArturoQuintana committed:
Minor modifications following PEP8 convention related to comments formats and uses of parentheses

ArturoQuintana committed:
Decoupling the test setting process from the SetUp function inherited from TestCase. This will allow using different topologies over the same test class, enhancing tests extensibility.

ArturoQuintana committed:
Minor modifications following PEP8 convention related to comment formats and whitespaces

ArturoQuintana committed:
Removing unused variables

ArturoQuintana committed:
Modification on the modules access path on importing

ArturoQuintana committed:
Inclusion of new packages to support testing extension with the implementation of new resources related to the topology specification

ArturoQuintana committed:
Inclusion of new resources related to the topology specification to support new tests

ArturoQuintana committed:
Improved the exception handling to cover the case a node can not be added to the topology graph

ArturoQuintana committed:
Elimination of the _set_default_metadata function. It breaks the pathfinding process due to the inclusion of empty metadata in all the links

ArturoQuintana committed:
Minor modifications following PEP8 convention related to comment formats and whitespaces

ArturoQuintana committed:
- Minor modifications following PEP8 convention related to comment formats and whitespaces
- Inclusion of new package to support testing extension

ArturoQuintana committed:
Minor modifications following PEP8 convention related to comment formats and whitespaces

ArturoQuintana committed:
- Creation of new methods to support the testing process
- Optimized the dependencies between methods

ArturoQuintana committed:
- Splitting tests on simple assertions (one by a test)
- Creation of more tests to cover a wider range of code

ArturoQuintana committed:
- Cleaning up the code

ArturoQuintana committed:
- Splitting tests on simple assertions (one by a test)
- Creation of more tests to cover a wider range of code
- Creation of new methods to support the testing process
- Optimized the dependencies between methods

ArturoQuintana committed:
- Cleaning up the code

ArturoQuintana committed:
- Splitting tests on simple assertions (one by a test)

ArturoQuintana committed:
- Cleaning up the code

ArturoQuintana committed:
- Inclusion of new test to address more possible exceptions

ArturoQuintana committed:
Handling more exceptions to address a greater chance of failures

ArturoQuintana committed:
Handling more exceptions to address a greater chance of failures

ArturoQuintana committed:
Following Pylint recommendations

ArturoQuintana committed:
Change requirements to always install latest version of exactdelaypathfinder

ArturoQuintana committed:
Fixing some misspelling in the

ArturoQuintana committed:
Refactoring to reduce duplicate code

ArturoQuintana committed:
Refactoring to reduce duplicate code

ArturoQuintana committed:
Setting some variables

ArturoQuintana committed:
Eliminate unnecessary initializations

ArturoQuintana committed:
Cleaning up the code

ArturoQuintana committed:
Breaking multiple asserts into independent methods

ArturoQuintana committed:
Cleaning up the code

ArturoQuintana committed:
Extending the code for reusability

ArturoQuintana committed:
Breaking multiple asserts into independent methods

ArturoQuintana committed:
Cleaning up the code

ArturoQuintana committed:
Breaking multiple asserts into independent methods

ArturoQuintana committed:
Cleaning up the code

ArturoQuintana committed:
Breaking multiple asserts into independent methods

ArturoQuintana committed:
Cleaning up the code

josemauro commented:
Hey @ArturoQuintana, thank you for this pull request. It will be very helpful if you simplify the PR text. You can use the original title for each section in the PR. You can write a text for each section following the description on the template. If you think that a specific section is not necessary, you can remove it. If the PR are not ready to review yet, you can change it for a draft PR. I hope this tips help you reach a more simple text.

ArturoQuintana committed:
Breaking multiple asserts into independent methods

ArturoQuintana committed:
Cleaning up the code

ArturoQuintana committed:
Breaking multiple asserts into independent methods

ArturoQuintana committed:
Cleaning up the code

ArturoQuintana committed:
Following Pylint recommendations

ArturoQuintana committed:
- Following Pylint recommendations
- Cleaning up the code

ArturoQuintana committed:
- Following Pylint recommendations

ArturoQuintana committed:
- Minor changes

ArturoQuintana committed:
- Extracting to a different file those methods that support the testing process

ArturoQuintana committed:
- Extracting to a different file those methods that support the testing process (Edges settings related)

ArturoQuintana committed:
Following pylint recommendation

ArturoQuintana committed:
Extracting to a different file those methods that support the testing process (Edges settings related)

ArturoQuintana committed:
Splitting a big test's file into different files to tackle a small part of the tests

ArturoQuintana committed:
Minor changes

ajoaoff commented on graph.py:
```diff
@@ -75,8 +133,76 @@ def shortest_paths(self, source, destination, parameter=None):
+                if paths:
+                    results.append(
+                        {"paths": paths, "metrics": {**base, **additional}})
+            i = i + 1
```
This is using combinations with sizes from `length-minimum_hits` to `length`. 
Shouldn't it use combinations from `minimum_hits` to `length`?
```suggestion
        i = minimum_hits
        while paths == [] and i <= length:
            for combo in combinations(flexible.items(), i):
                additional = dict(combo)
                paths = self._constrained_shortest_paths(
                    source, destination,
                    self._filter_links(first_pass_links,
                                       metadata=False, **additional))
                if paths:
                    results.append(
                        {"paths": paths, "metrics": {**base, **additional}})
            i = i + 1
```

ArturoQuintana replied:
Changes applied!
Thanks!

ajoaoff commented on graph.py:
```diff
@@ -46,23 +100,27 @@ def update_links(self, links):
                 self.graph.add_edge(link.endpoint_a.id, link.endpoint_b.id)
                 for key, value in link.metadata.items():
-                    keys.append(key)
+                    keys.append(key) if key not in keys else keys
```
`keys` is not used anymore, as `_set_default_metadata` is not called.

ArturoQuintana replied:
Changes applied!
Thanks!

ArturoQuintana committed:
Following changes recommendations

ArturoQuintana committed:
Cleaning up the code

ArturoQuintana committed:
- Modification of the minimum_hits variable in case the user sets a minimum greater than the length
- Modification on the implementation of the constrained_flexible_paths to avoid reverse order results

ArturoQuintana committed:
- Cleaning up the code
